wasm/wat/convert_test.go

```diff
@@ -0,0 +1,214 @@
+package wat
+
+import (
+	"fmt"
+	"os"
+	"testing"
+	"unicode/utf8"
+
+	"github.com/bytecodealliance/wasmtime-go"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/wazero/wasm"
+)
+
+func TestTextToBinary(t *testing.T) {
+	zero := uint32(0)
+	tests := []struct {
+		name     string
+		input    string
+		expected *wasm.Module
+	}{
+		{
+			name:     "empty",
+			input:    "(module)",
+			expected: &wasm.Module{},
+		},
+		{
+			name:  "import func empty",
+			input: "(module (import \"foo\" \"bar\" (func)))", // ok empty sig
+			expected: &wasm.Module{
+				TypeSection: []*wasm.FunctionType{{}},
+				ImportSection: []*wasm.ImportSegment{{
+					Module: "foo", Name: "bar",
+					Desc: &wasm.ImportDesc{
+						Kind:          wasm.ImportKindFunction,
+						FuncTypeIndex: 0,
+					},
+				}},
+			},
+		},
+		{
+			name:  "import func two",
+			input: "(module (import \"foo\" \"bar\" (func)) (import \"baz\" \"qux\" (func)))",
+			expected: &wasm.Module{
+				TypeSection: []*wasm.FunctionType{{}},
+				ImportSection: []*wasm.ImportSegment{
+					{
+						Module: "foo", Name: "bar",
+						Desc: &wasm.ImportDesc{
+							Kind:          wasm.ImportKindFunction,
+							FuncTypeIndex: 0,
+						},
+					}, {
+						Module: "baz", Name: "qux",
+						Desc: &wasm.ImportDesc{
+							Kind:          wasm.ImportKindFunction,
+							FuncTypeIndex: 0,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "start imported function by name",
+			input: `(module
+	(import "" "hello" (func $hello))
+	(start $hello)
+)`,
+			expected: &wasm.Module{
+				TypeSection: []*wasm.FunctionType{{}},
+				ImportSection: []*wasm.ImportSegment{{
+					Module: "", Name: "hello",
+					Desc: &wasm.ImportDesc{
+						Kind:          wasm.ImportKindFunction,
+						FuncTypeIndex: 0,
+					},
+				}},
+				StartSection: &zero,
+			},
+		},
+		{
+			name: "start imported function by index",
+			input: `(module
+	(import "" "hello" (func))
+	(start 0)
+)`,
+			expected: &wasm.Module{
+				TypeSection: []*wasm.FunctionType{{}},
+				ImportSection: []*wasm.ImportSegment{{
+					Module: "", Name: "hello",
+					Desc: &wasm.ImportDesc{
+						Kind:          wasm.ImportKindFunction,
+						FuncTypeIndex: 0,
+					},
+				}},
+				StartSection: &zero,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			m, err := TextToBinary([]byte(tc.input))
+			require.NoError(t, err)
+			require.Equal(t, tc.expected, m)
+		})
+	}
+}
+
+func TestTextToBinary_Errors(t *testing.T) {
+	tests := []struct{ name, input, expectedErr string }{
+		{
+			name:        "start, but no funcs",
+			input:       "(module (start $main))",
+			expectedErr: "1:16: unknown function name: $main in module.start",
```

Comment:
First time using line/col outside the scope of the tokenParser function

Reply:
maybe we have `${filename}:${line}:${col}` format as that is recognizable by some editors and we could just click it to jump to the position (e.g. VSCode)

Reply:
should

Reply:
ah - filename does not exist in this context!

Reply:
nvm

Reply:
yeah that's why I exported `FormatError`, so that the formatting can be changed even if the default formatting could be ok for inlined text (which doesn't have a filename)